### ISSUE TITLE
Feature request: gradle plugin support for android library projects

### ISSUE THREAD

**Issue by amir169** [getsentry/sentry-java#673]
We want to use sentry in an android library project. We followed this documentation <https://docs.sentry.io/clients/java/modules/android/> and it seems to work fine if we just init Sentry inside our library but we failed to use gradle plugin for Proguard files.

Is there any way to send Proguard mapping of an android library to Sentry?

**Comment by bretthoerner** [comment]
Hmm, I'm not sure this is possible. The end user(s) will Proguard their application which will in turn generate unique mappings of their application and all library code, right? So when the crash occurs at runtime, their application code may have a method named `fg` that is really your method `bg`.

Do you have any ideas of how to get around that?

**Comment by amir169** [comment]
Yes. There is a way to prevent your classes from being obfuscated by end users.
Using the property ```consumerProguardFiles ${rules}``` of Android Gradle DSL, you can make your users to respect your _rules_. 
So you can obfuscate your own _classes_ and add a ```-keep ${classes}``` rule in consumer _rules_ and the user will not obfuscate your _classes_ anymore.

[Here](https://proandroiddev.com/handling-proguard-as-library-developer-or-in-a-multi-module-android-application-2d738c37890) is an explanation about ```consumerProguardFiles```

**Comment by ninniuz** [comment]
@amir169 what is your use case? Are you minifying your library? If yes, please note that the mapping file for your library will differ from the one generated for a minified app which uses your library, resulting in unretraceable stracktraces.

I have configured Sentry in a library project as well, which is used as a dependency in different apps. Sentry config is applied on each of the apps though, not the library itself.

**Comment by amir169** [comment]
@ninniuz 
As I explained in previous comment, you can make your users to _keep_ some specified classes using _consumerProguardFiles_. For example I repackage all my internal library classes to a package named _internalPackage_ and put a proguard _keep_ rule in consumer proguard file to keep _internalPackage_. In this way I can prevent the library user from obfuscating my _internalPackage_.

I agree I can't de-obfuscate all stack traces but stack traces which contain _internalPackage_ classes can be de-obfuscated because they are not changed by library user.

I don't like to force my users to do extra step by adding Sentry gradle plugin.

**Comment by ninniuz** [comment]
Well if you add keep rules for your package then your classes won't be obfuscated, hence there will be no need to retrace them nor to upload any ProGuard mapping file to sentry at all :-)

**Comment by bretthoerner** [comment]
I assume he needs his library to be minified for users.

@amir169 Back to the original question,

> it seems to work fine if we just init Sentry inside our library but we failed to use gradle plugin for Proguard files

If you're going to tell your users to not re-minify your library, and if you are the one creating and controlling your own Sentry client, then I think what you'll need is a different way to provide the UUID that matches the current library version to the debug symbols in Sentry:

https://github.com/getsentry/sentry-java/blob/d333996cfc45499e0eb96e96c1cbb9e37eff1971/sentry-android/src/main/java/io/sentry/android/event/helper/AndroidEventBuilderHelper.java#L152-L177

It's currently hardcoded to look for a properties file that our plugin adds to the app. We could make it `protected` and you could use your own subclass that finds/hardcodes the UUID elsewhere?

**Comment by amir169** [comment]
@bretthoerner 
Yes. I minify my library and make my users not to re-minify. Sorry for ambiguity. Thank you.

So in that case, what should I do?
I should use _sentry-cli_ to upload Proguard and get UUID(s) and then return that UUID(s) using _getProGuardUuids_ method. Am I right?

If so, is there a way to inject my custom _AndroidEventBuilderHelper_ to the rest of library?
Is there a plan to make _getProGuardUuids_ protected?

Thanks

**Comment by bretthoerner** [comment]
You could still use our Gradle plugin (or upload manually if you wish) but then you need to take the UUID from the `sentry-debug-meta.properties` we drop in the resources and somehow make it available to your library via `getProGuardUuids`, correct.

**Comment by amir169** [comment]
OK

Let say I extended ```AndroidEventBuilderHelper```. How can I use it? Is there any documentation?

Another question is:
When the change (making ```getProGuardUuids``` protected) will be done?
